pkg/rdt/rdt.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2019 Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rdt
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	"github.com/intel/cri-resource-manager/pkg/cri/resource-manager/cache"
+	logger "github.com/intel/cri-resource-manager/pkg/log"
+	"github.com/intel/cri-resource-manager/pkg/utils"
+)
+
+const resctrlGroupPrefix = "cri-resmgr."
+
+// RdtControl is the interface managing Intel RDT resources
+type RdtControl interface {
+	// GetClasses returns the names of RDT classes (or resctrl control groups)
+	// available
+	GetClasses() []string
+
+	//
+	SetClassPID(string, string) error
+
+	SetClassContainer(string, cache.Container) error
+
+	// SetConfig re-configures RDT resources
+	SetConfig(string) error
+}
+
+type rdtControl struct {
+	logger.Logger
+
+	info rdtInfo
+	conf RdtConfig
+}
+
+type RdtConfig struct {
+	ResctrlGroups map[string]ResctrlGroupConfig `json:"resctrlGroups"`
+}
+
+func NewRdtControl(resctrlpath string, config string) (RdtControl, error) {
+	var err error
+	r := &rdtControl{Logger: logger.NewLogger("rdt")}
+
+	// Get info from the resctrl filesystem
+	r.info, err = getRdtInfo(resctrlpath)
+	if err != nil {
+		return nil, err
+	}
+	RdtCacheBitmaskNumBits = uint64(r.info.l3.cbmMask.lsZero())
+
+	// Configure resctrl
+	r.conf, err = parseConfData([]byte(config))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.configureResctrl(r.conf); err != nil {
+		return nil, rdtError("configuration failed: %v", err)
+	}
+
+	return r, nil
+}
+
+func (r *rdtControl) GetClasses() []string {
+	ret := make([]string, len(r.conf.ResctrlGroups))
+
+	i := 0
+	for k, _ := range r.conf.ResctrlGroups {
+		ret[i] = k
+		i++
+	}
+	sort.Strings(ret)
+	return ret
+}
+
+func (r *rdtControl) SetClassContainer(class string, container cache.Container) error {
+	cId := container.GetId()
+	pod, ok := container.GetPod()
+	if !ok {
+		return rdtError("Pod of container %q not found", cId)
+	}
+	cgroupParent := pod.GetCgroupParentDir()
+
+	pids, err := utils.GetProcessInContainer(cgroupParent, cId)
+	if err != nil {
+		return rdtError("failed to get PIDs of container %s: %v", cId, err)
+	}
+
+	for _, pid := range pids {
+		if err = r.SetClassPID(class, strconv.Itoa(pid)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *rdtControl) SetClassPID(class, pid string) error {
```

Comment:
nitpick: IMO it is more natural to say that the `RDT class` is a property of the `process` and not the other way around. To back this up/or to look at it in another way: setting the class of a process does not/shpuld not change the class of other processes, but `SetClassPID`, at least to me, seems to suggest a one-to-one relationship between the two, thus soft-implying that setting the class pid would undo a previously set one.

Therefore, I'd rename this to `SetPIDClass` or preferably to the more descriptive `SetProcessClass`.

Reply:
Agree, the naming was bad. Changed as you suggested